Added MitsubishiAC VaneLeft

hideosasaki opened the pull request:
Hi,
I am a user of Mitsubishi A/C MSZ-ZW4017S-W, which has two vanes. While evaluating the library I noticed that IRMitsubishiAC::setVane moved only the right vane. So I added setVaneLeft method. 
This is the first time to contribute open-source projects for me. Thank you so much for correcting my mistakes.
Hideo
![IMG_0284](https://user-images.githubusercontent.com/2324488/130717720-9a22f3bb-daeb-4def-aa1e-377ca9c0769f.jpg)

hideosasaki committed:
Added MitsubishiAC VaneLeft

NiKiZe commented:
Nice start! ありがとうございます!

Just a quick initial look.
Please add your device to the Supports section in the `.h` file.
There is a few checks that is failing, can you take a look?

Maybe we can also add some captures from original remote and add tests?
There is also some string generation that we would like to have. Let us know if this is something that you want to tackle, and if you need guidance. Otherwise we are happy to do these parts before merging.

NiKiZe commented on src/ir_Mitsubishi.h:
```diff
@@ -271,6 +279,8 @@ class IRMitsubishiAC {
   uint8_t getVane(void) const;
   uint8_t getWideVane(void) const;
+  void setVaneLeft(const uint8_t position);
+  uint8_t getVaneLeft(void) const
```
Missing `;`

NiKiZe commented on src/ir_Mitsubishi.h:
```diff
@@ -84,8 +84,16 @@ union Mitsubishi144Protocol{
+    uint8_t          :4;
+    // Byte 15
+    uint8_t          :4;
+    uint8_t          :4;
```
For now these can be kept as `:8;` instead of double? `:4;` or is there some specific reason that it is split?

NiKiZe commented on src/ir_Mitsubishi.cpp:
```diff
@@ -480,6 +480,19 @@ uint8_t IRMitsubishiAC::getWideVane(void) const {
+uint8_t IRMitsubishiAC::getVaneLeft(void) const {
+  return _.VaneLeft;
+}
+
```
I think this fits in 80 columns on one line?

NiKiZe commented on src/ir_Mitsubishi.cpp:
```diff
@@ -480,6 +480,19 @@ uint8_t IRMitsubishiAC::getWideVane(void) const {
+/// @param[in] position The position/mode to set the vane to.
+void IRMitsubishiAC::setVaneLeft(const uint8_t position) {
+  uint8_t pos = std::min(position, kMitsubishiAcVaneAutoMove);  // bounds check
+  _.VaneLeft = pos;
```
Temporary `pos` variable is unnecessary

hideosasaki committed:
refactored and fixed typo

hideosasaki commented:
@NiKiZe 
Thank you for reviewing! I committed some fixes you pointed out. And I am sorry for not being able to provide the real capture, because I haven't got any IR receivable devices but only have M5Stick-C.

crankyoldgit commented on test/ir_Mitsubishi_test.cpp:
```diff
@@ -652,6 +652,26 @@ TEST(TestMitsubishiACClass, VaneMode) {
   EXPECT_EQ(kMitsubishiAcVaneAutoMove - 1, ac.getVane());
 }
 
+TEST(TestMitsubishiACClass, VaneMode) {
```
A test with the same name already exists. e.g. `VaneMode`
You need to change it for your new test.
e.g.
```c++
TEST(TestMitsubishiACClass, VaneLeft) {
```

hideosasaki replied:
Changed to VaneLeft

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -480,6 +480,16 @@ uint8_t IRMitsubishiAC::getWideVane(void) const {
   return _.WideVane;
 }
 
+/// Set the requested Left Vane (Vertical Swing) operation mode of the a/c unit.
```
Is it "Vertical Swing" or "Horizontal Swing"? Your description `Mitsubishi144Protocol` (`VaneLeft`) indicates it's a "SwingH(Left)". Hence the question.

hideosasaki replied:
it's Vertical Swing. The comment was wrong. See the attached photo on the first post.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -480,6 +480,16 @@ uint8_t IRMitsubishiAC::getWideVane(void) const {
+  _.VaneLeft = std::min(position, kMitsubishiAcVaneAutoMove);  // bounds check
+}
+
+/// Get the Left Vane (Vertical Swing) mode of the A/C.
```
Ditto other comment.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -480,6 +480,16 @@ uint8_t IRMitsubishiAC::getWideVane(void) const {
+/// Set the requested Left Vane (Vertical Swing) operation mode of the a/c unit.
+/// @param[in] position The position/mode to set the vane to.
+void IRMitsubishiAC::setVaneLeft(const uint8_t position) {
+  _.VaneLeft = std::min(position, kMitsubishiAcVaneAutoMove);  // bounds check
```
[Question] Are all the possible values the same as for `setVane()` etc?

hideosasaki replied:
I believe so. My real remote control has the same settings both left and right.

crankyoldgit reviewed (changes requested):
Congrats on your effort. We appreciate it!

hideosasaki committed:
fixed duplicated name

hideosasaki committed:
fixed comment

crankyoldgit commented:
One of the unit tests is failing.
It just needs to be updated.
See https://github.com/crankyoldgit/IRremoteESP8266/pull/1572/checks?check_run_id=3419003709#step:4:2931

hideosasaki committed:
excepted VaneLeft from the IR message test

hideosasaki commented:
@crankyoldgit 
I fixed the test case, just removed the VaneLeft setting from. I had no choice to do the bad thing because I haven't got the IR receivable device.

crankyoldgit commented on test/ir_Mitsubishi_test.cpp:
```diff
@@ -725,7 +725,6 @@ TEST(TestMitsubishiACClass, MessageConstuction) {
   EXPECT_EQ(kMitsubishiAcCool, ac.getMode());
   EXPECT_EQ(27, ac.getTemp());
   EXPECT_EQ(3, ac.getVane());
-  EXPECT_EQ(2, ac.getVaneLeft());
```
That's not going to fix it. Put back the line. It's not failing there.
It's failing on the follow test a few lines down. e.g:
```c++
EXPECT_EQ(
      "f38000d50"
      "m3400s1750"
      ...
```

The reason it is failing is because add/changing the VaneLeft value, it is producing a slightly different message.
Just run the test, grab the generated output, and replace the "expected" output with that (properly/nicely formatted like the existing one)

The resulting diff will only have probably about two lines different. That's due to the changed bits in the message.

i.e. The existing data in the test won't match the new message you've created, so you need to update it.

hideosasaki replied:
Thank you for the nice explanation. I committed the updated test code.

hideosasaki committed:
updated an assert data changed by adding VaneLeft

crankyoldgit reviewed (commented):
Approved

crankyoldgit commented:
FYI, the changes mentioned above have been included in the just released v2.7.20 of the library.
